src/instructions/instruction.njk

```diff
@@ -18,22 +18,124 @@ eleventyComputed:
 
 <section>
   <h2>Encoding</h2>
+
   <p><strong>Type:</strong> {{ inst.encodingType or "" }}</p>
-  <div class="encoding">
-    <div class="bitfield">
-      {% set fields = inst.encoding.fields %}
-      {% for f in fields %}
-        <div class="field {{ f.kind }}" style="flex: {{ f.width }}">
-          <div class="label">{{ f.label }}</div>
-          <div class="range">bits {{ f.from }}..{{ f.to }}</div>
-        </div>
-      {% endfor %}
-    </div>
-    <div class="bits">
-      {% if inst.encoding.funct7 %}<div><strong>funct7</strong>: {{ inst.encoding.funct7 }}</div>{% endif %}
-      {% if inst.encoding.funct3 %}<div><strong>funct3</strong>: {{ inst.encoding.funct3 }}</div>{% endif %}
-      {% if inst.encoding.opcode %}<div><strong>opcode</strong>: {{ inst.encoding.opcode }}</div>{% endif %}
-    </div>
+
+  <div class="bitfield-diagram">
+    <script type="text/bitfield">
+    {{ inst.bitfieldJSON | dump | safe }}
+    </script>
+  </div>
+
+  <script>
+  (function () {
+    function jsonmlToDom(node, inSvg) {
+      if (node === null || node === undefined) return null;
+      if (typeof node === "string" || typeof node === "number") {
+        return document.createTextNode(String(node));
+      }
+      if (!Array.isArray(node) || node.length === 0) return null;
+
+      const [tagName, ...rest] = node;
+      let attrs = {};
+      let children = rest;
+
+      if (children.length && typeof children[0] === "object" && !Array.isArray(children[0])) {
+        attrs = children.shift() || {};
+      }
+
+      const isSvgNode = inSvg || tagName === "svg";
+      const element = isSvgNode
+        ? document.createElementNS("http://www.w3.org/2000/svg", tagName)
+        : document.createElement(tagName);
+
+      Object.entries(attrs).forEach(([key, value]) => {
+        element.setAttribute(key, value);
+      });
+
+      children.forEach(child => {
+        const childNode = jsonmlToDom(child, isSvgNode);
+        if (childNode) element.appendChild(childNode);
+      });
+
+      return element;
+    }
+
+    function parseBitfieldData(raw) {
+      if (!raw) return null;
+      try {
+        const payload = JSON.parse(raw);
+        if (Array.isArray(payload)) {
+          return { segments: payload };
+        }
+        if (payload && Array.isArray(payload.reg)) {
+          return { segments: payload.reg, options: payload.options };
+        }
+        if (payload && Array.isArray(payload.fields)) {
+          return { segments: payload.fields, options: payload.options };
+        }
+      } catch (err) {
+        console.warn("Failed to parse bitfield JSON", err);
+      }
+      return null;
+    }
+
+    function renderBitfieldScripts() {
+      const scripts = document.querySelectorAll('script[type="text/bitfield"]');
```

Comment:
I'm a tiny bit confused about the overall strategy here. Maybe it just needs more comments? I think what you're doing is finding a way to embed the underlying JSON data in a way that the frontend will then parse? And the `<script>` tag is a hack for embedding that JSON string?

A little bit of commenting here could go a long way!